pkg/operator/environmentchecker/environmentcheckcontroller.go

```diff
@@ -0,0 +1,409 @@
+	apiClients               utils.ApiClients
+	controllers              []conditionalController
+	storageClassController   *storageclasscontroller.StorageClassController
+	operandControllerStarted bool
+	vSphereConnection        *vclib.VSphereConnection
+	vSphereChecker           vSphereEnvironmentCheckInterface
+}
+
+const (
+	cloudConfigNamespace              = "openshift-config"
+	infraGlobalName                   = "cluster"
+	secretName                        = "vmware-vsphere-cloud-credentials"
+	defaultNamespace                  = "openshift-cluster-csi-drivers"
+	driverOperandName                 = "vmware-vsphere-csi-driver"
+	resyncDuration                    = 20 * time.Minute
+	envVMWareVsphereDriverSyncerImage = "VMWARE_VSPHERE_SYNCER_IMAGE"
+	storageClassControllerName        = "VMwareVSphereDriverStorageClassController"
+	storageClassName                  = "thin-csi"
+)
+
+type conditionalControllerInterface interface {
+	Run(ctx context.Context, workers int)
+}
+
+type conditionalController struct {
+	name       string
+	controller conditionalControllerInterface
+}
+
+func NewEnvironmentCheckController(
+	name, targetNamespace string,
+	apiClients utils.ApiClients,
+	recorder events.Recorder,
+) factory.Controller {
+	kubeInformers := apiClients.KubeInformers
+	ocpConfigInformer := apiClients.ConfigInformers
+	configMapInformer := kubeInformers.InformersFor(cloudConfigNamespace).Core().V1().ConfigMaps()
+	infraInformer := ocpConfigInformer.Config().V1().Infrastructures()
+	scInformer := kubeInformers.InformersFor("").Storage().V1().StorageClasses()
+	csiDriverLister := kubeInformers.InformersFor("").Storage().V1().CSIDrivers().Lister()
+	csiNodeLister := kubeInformers.InformersFor("").Storage().V1().CSINodes().Lister()
+	nodeLister := apiClients.NodeInformer.Lister()
+
+	rc := recorder.WithComponentSuffix("vmware-" + strings.ToLower(name))
+
+	c := &EnvironmentCheckController{
+		name:            name,
+		targetNamespace: targetNamespace,
+		kubeClient:      apiClients.KubeClient,
+		operatorClient:  apiClients.OperatorClient,
+		configMapLister: configMapInformer.Lister(),
+		secretLister:    apiClients.SecretInformer.Lister(),
+		csiNodeLister:   csiNodeLister,
+		scLister:        scInformer.Lister(),
+		csiDriverLister: csiDriverLister,
+		nodeLister:      nodeLister,
+		apiClients:      apiClients,
+		eventRecorder:   rc,
+		vSphereChecker:  newVSphereEnvironmentChecker(),
+		infraLister:     infraInformer.Lister(),
+	}
+	c.controllers = []conditionalController{}
+	c.createCSIDriver()
+	c.storageClassController = c.createStorageClassController()
+
+	return factory.New().WithInformers(
+		apiClients.OperatorClient.Informer(),
+		configMapInformer.Informer(),
+		apiClients.SecretInformer.Informer(),
+		infraInformer.Informer(),
+		scInformer.Informer(),
+	).WithSync(c.sync).
+		ResyncEvery(resyncDuration).
+		WithSyncDegradedOnError(apiClients.OperatorClient).ToController(c.name, rc)
+}
```

Comment:
I'd prefer this controller to stick to checking the environment only, running side-by-side with other controllers. Currently it acts like a manger and IMO it's getting too complex.

If this controller needs to share some of its findings to other controllers, I'd suggest using another custom resource for that (e.g., the ClusterCSIDriver CR).

Reply:
We discussed this offline. We can't start CSI controller set controllers if existing driver check failed (otherwise we will have two CSI drivers). Also if we are going to put something in  `ClusterCSIDriver` object for conditional start of CSI controller set, then we will have to wrap CSI controller set in another controller and at that point, I am not sure if this is superior or current approach.

Reply:
We did something like this in Manila: the `ManilaController` does a few checks, starts the remaining controllers and sets a `*Disabled` condition. Later on CSO [deals with that condition](https://github.com/bertinatto/cluster-storage-operator/blob/2e76e02cb8479784a582e57a587be8fd5d75cae2/pkg/operator/csidriveroperator/crcontroller.go#L192-L195).

Should we do the same here?